core/sdk-core/src/main/java/software/amazon/awssdk/core/async/AsyncRequestBodyFromInputStreamConfiguration.java

```diff
@@ -0,0 +1,200 @@
+    }
+
+    /**
+     * @return the provided {@link ExecutorService}.
+     */
+    public ExecutorService executor() {
+        return executor;
+    }
+
+    /**
+     * @return the provided max read limit used to mark and reset the {@link InputStream}).
+     */
+    public Integer maxReadLimit() {
+        return maxReadLimit;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        AsyncRequestBodyFromInputStreamConfiguration that = (AsyncRequestBodyFromInputStreamConfiguration) o;
+
+        if (!Objects.equals(inputStream, that.inputStream)) {
+            return false;
+        }
+        if (!Objects.equals(contentLength, that.contentLength)) {
+            return false;
+        }
+        if (!Objects.equals(executor, that.executor)) {
+            return false;
+        }
+        return Objects.equals(maxReadLimit, that.maxReadLimit);
+    }
+
+    @Override
+    public int hashCode() {
+        int result = inputStream != null ? inputStream.hashCode() : 0;
+        result = 31 * result + (contentLength != null ? contentLength.hashCode() : 0);
+        result = 31 * result + (executor != null ? executor.hashCode() : 0);
+        result = 31 * result + (maxReadLimit != null ? maxReadLimit.hashCode() : 0);
+        return result;
+    }
+
+    /**
+     * Create a {@link Builder}, used to create a {@link AsyncRequestBodyFromInputStreamConfiguration}.
+     */
+    public static Builder builder() {
+        return new DefaultBuilder();
+    }
+
+
+    @Override
+    public AsyncRequestBodyFromInputStreamConfiguration.Builder toBuilder() {
+        return new DefaultBuilder(this);
+    }
+
+    public interface Builder extends CopyableBuilder<AsyncRequestBodyFromInputStreamConfiguration.Builder,
+        AsyncRequestBodyFromInputStreamConfiguration>  {
+
+        /**
+         * Configures the InputStream.
+         *
+         * @param inputStream the InputStream
+         * @return This object for method chaining.
+         */
+        Builder inputStream(InputStream inputStream);
+
+        /**
+         * Configures the length of the provided {@link InputStream}
+         * @param contentLength the content length
+         * @return This object for method chaining.
+         */
+        Builder contentLength(Long contentLength);
+
+        /**
+         * Configures the {@link ExecutorService} to perform the blocking data reads.
+         *
+         * @param executor the executor
+         * @return This object for method chaining.
+         */
+        Builder executor(ExecutorService executor);
+
+        /**
+         * Configures max read limit used to mark and reset the {@link InputStream}. This will have no
+         * effect if the stream doesn't support mark and reset.
+         *
+         * <p>
+         * By default, it is 128 KiB.
```

Comment:
Alternatively, we could make it contentLength + 1 so that users don't have to configure it by themselves, however, this would mean the SDK will buffer all the data, so I'm reluctant to make this the default behavior.